frontend/appflowy_tauri/src/appflowy_app/utils/slate/text.ts

```diff
@@ -0,0 +1,30 @@
+import { Editor, Element, Text, Location } from 'slate';
+import { TextDelta } from '$app/interfaces/document';
+
+export function getDelta(editor: Editor, at: Location): TextDelta[] {
+  const baseElement = Editor.fragment(editor, at)[0] as Element;
+  return baseElement.children.map((item) => {
+    const { text, ...attributes } = item as Text;
+    return {
+      insert: text,
+      attributes,
+    };
+  });
+}
+
+export function getRetainRangeBy(editor: Editor) {
+  const start = Editor.start(editor, editor.selection!);
+  return {
+    anchor: { path: [0, 0], offset: 0 },
+    focus: start,
+  };
+}
+
+export function getInsertRangeBy(editor: Editor) {
+  const end = Editor.end(editor, editor.selection!);
+  const fragment = (editor.children[0] as Element).children;
+  return {
+    anchor: end,
+    focus: { path: [0, fragment.length - 1], offset: (fragment[fragment.length - 1] as Text).text.length },
```

Comment:
`(fragment[fragment.length - 1] as Text).text.length`

This line seems a bit complex.

```ts
const last = fragment[fragment.length - 1] // By the way, it looks like the fragment variable is an array or a list. Can we get the last value of the array using `fragment.last`?
const offset = last ? last.text.length : 0; 
```

Reply:
Maybe we just do it:
const lastIndex = fragment.length - 1;
//...
focus: { path: [0, lastIndex], offset: (fragment[lastIndex] as Text).text.length },